### PR TITLE
fix negative impls inference

### LINKED ISSUE #74383: negative impls cause problems with type inference
```rust
#![feature(negative_impls)]

struct Foo<T>(T);

impl !Send for Foo<()> {}

fn test<T>() -> T where Foo<T>: Send { todo!() }

fn main() {
    let _: u8 = test();
}
```
Fails with
```
error[E0277]: `Foo<_>` cannot be sent between threads safely
  --> $DIR/fk.rs:10:17
   |
LL | fn test<T>() -> T where Foo<T>: Send { todo!() }
   |                                 ---- required by this bound in `test`
...
LL |     let _: u8 = test();
   |                 ^^^^ `Foo<_>` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `Foo<_>`
```
Explicitly specifying `test::<u8>()` works though.

### PR DESCRIPTION
**Pull request by lcnr**
fixes #74383

r? @nikomatsakis @matthewjasper

### CODE REVIEW

File: src/librustc_trait_selection/traits/select/mod.rs [hunk 1]
```diff
@@ -934,16 +934,34 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
...
+    /// interpreted as ambiguity.
     fn filter_negative_and_reservation_impls(
         &mut self,
+        pred: ty::PolyTraitPredicate<'tcx>,
         candidate: SelectionCandidate<'tcx>,
     ) -> SelectionResult<'tcx, SelectionCandidate<'tcx>> {
         if let ImplCandidate(def_id) = candidate {
             let tcx = self.tcx();
             match tcx.impl_polarity(def_id) {
                 ty::ImplPolarity::Negative if !self.allow_negative_impls => {
```
Comment (lcnr) [inline src/librustc_trait_selection/traits/select/mod.rs:950]: So we end up here if we have exactly one impl candidate which is negative.

We only have a negative impl for `Foo<()>` here, while the self type is actually more generic (`Foo<_>`).
So if `Foo` is an auto trait we should return ambiguous in this case, as `Foo` can be implemented here, e.g. `Foo<u8>`.

I don't know how we can detect that the impl is at least as general as the current self type rn though, need some help for this. (I don't think I have to write a new type relation for this :thinking: )

Comment (nikomatsakis) [reply src/librustc_trait_selection/traits/select/mod.rs:950]: Ah, interesting.

Comment (nikomatsakis) [reply src/librustc_trait_selection/traits/select/mod.rs:950]: So actually I'm not sure there is a bug here. There has been a long-standing debate about the correct semantics of auto traits in the face of negative impls, but the current semantics (not, perhaps *100%* settled) is that

* If the user writes ANY explicit impl, positive or negative, then the compiler generates NO automatic impl

This was perhaps to be paired with disallowing negative impls that were overly narrow, such as `!Foo for Bar<()>`. 

So in this case, writing `impl !Send for Bar<()>` (for example) would mean that the compiler does not add an automatic `Send for Bar`  impl at all.

It's debatable what's more intuitive. I personally prefer the current rules -- I would prefer that if you write *anything* explicit, you must write *everything* explicit. But also this avoids having a setup (which specialization/coherence currently do not permit) where you have a "base impl" that accepts lots of things and then "negative impls" that cut out "exceptions" where that base impl no longer applies.

In other words, given specialization (and auto traits as currently implemented and specified), if you know that a "base impl" applies, you know that the trait is implemented, and no specialization can make that "untrue".

Comment (lcnr) [reply src/librustc_trait_selection/traits/select/mod.rs:950]: hmm, I think this makes sense :thinking: 

> disallowing negative impls that were overly narrow, such as !Foo for Bar<()>.

While negative impls are still unstable, this seems like a restriction we should probably enforce automatically.

Quite interestingly (or much rather, a logical consequence of how this is implemented rn), we can currently "fix" this inference issue by adding another nonsensical impl:
```rust
#![feature(negative_impls)]

struct Foo<T>(T);

struct IDontCare;

impl !Send for Foo<()> {}
unsafe impl Send for Foo<IDontCare> {}

fn test<T>() -> T where Foo<T>: Send { todo!() }

fn main() {
    let _: u8 = test();
}
```
I would love to use this to forbid calling `array_chunks::<0>()` in #74373 until we implement const where bounds,
but that may stop us from fixing negative impls (as we now need the broken version in std, even if only for an unstable method).

Comment (nikomatsakis) [reply src/librustc_trait_selection/traits/select/mod.rs:950]: > While negative impls are still unstable, this seems like a restriction we should probably enforce automatically.

these has been an open bug on this forever -- well, no, it's the first checkbox on https://github.com/rust-lang/rust/issues/13231. It'd be easy enough to do, we can use the same code we use for Drop impls.

Comment (lcnr) [reply src/librustc_trait_selection/traits/select/mod.rs:950]: thanks, might look into this in the following weeks :thinking: let's see how much time I have for this.
